src/ast/writing/writer.ts

```diff
@@ -63,3 +73,74 @@ export class ASTWriter {
         throw new Error("Expected an instance of ASTNode but got following: " + data);
     }
 }
+
+export class ASTSourceMapComputer {
+    private getSourceFragment(node: ASTNode, fragments: Map<ASTNode, string>): string {
+        const source = fragments.get(node);
+
+        if (source === undefined) {
+            /**
+             * @todo This happens for a few nodes that are present in ast by default.
+             *       The example would be a PararamterList node for a function with no returns.
+             *       They can not be written, so the result is empty string.
+             *       Better to clarify how we can handle such situations.
+             */
+            return "";
+        }
+
+        return source;
+    }
+
+    private getPrecomputedCoords(
+        node: ASTNode,
+        computed: Map<ASTNode, [number, number]>
+    ): [number, number] {
+        const coordinates = computed.get(node);
+
+        if (coordinates === undefined) {
+            throw new Error("Missing precomputed coordinates for the node " + node.print());
+        }
+
+        return coordinates;
+    }
+
+    private computeNodeCoords(
+        node: ASTNode,
+        fragments: Map<ASTNode, string>,
+        computed: Map<ASTNode, [number, number]>
+    ): [number, number] {
+        const parent = node.parent;
+        const sourceN = this.getSourceFragment(node, fragments);
+
+        if (parent === undefined) {
+            return [0, sourceN.length];
+        }
+
+        const sourceP = this.getSourceFragment(parent, fragments);
+        const [startP] = this.getPrecomputedCoords(parent, computed);
+
+        let offset = 0;
+
+        const sibling = node.previousSibling;
+
+        if (sibling) {
+            const [startS, lenS] = this.getPrecomputedCoords(sibling, computed);
+
+            offset = startS + lenS - startP;
+        }
+
+        const index = sourceP.indexOf(sourceN, offset);
```

Comment:
Why is this neccessary? Shouldn't it always be that `index == offset`?

Reply:
The `index` of the subnode is not the same as `offset` to start search. We also need to skip the whitespace and comments. Lets consider an example:

```solidity
contract A {
    function some() pure public {
        uint a = 1;
        // some comment
        uint b = 2;
    }
}
```

To pick location of `uint b = 2;`, we need to seach in source of the whole function definition. However, we need to skip the indentation space and comment here. The `offset` would replicate that we start search in function definition source just after the declaration `uint a = 1;` to skip collisions.

Reply:
Also added comments with clarification in https://github.com/ConsenSys/solc-typed-ast/pull/16/commits/0e89a4dc7faef6a2a63cb57d7568428e7db4d388.